File: app/ai-orchestrator-python/test/openai_compatible_smoke.py
import json
import os
import subprocess
import sys
import tempfile
import threading
import time
import urllib.request
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path


provider_calls = []


class FakeProvider(BaseHTTPRequestHandler):
    def do_POST(self):
        if self.path != "/chat/completions":
            self.send_response(404)
            self.end_headers()
            return
        length = int(self.headers.get("Content-Length", "0"))
        body = json.loads(self.rfile.read(length).decode("utf-8"))
        system = next((item.get("content", "") for item in body.get("messages", []) if item.get("role") == "system"), "")
        schema_name = "unknown"
        marker = "schema_name="
        if marker in system:
            schema_name = system.split(marker, 1)[1].split("，", 1)[0].split(",", 1)[0].split()[0]
        provider_calls.append(
            {
                "schema_name": schema_name,
                "authorization": self.headers.get("Authorization", ""),
                "response_format": body.get("response_format", {}).get("type"),
                "model": body.get("model"),
            }
        )
        payload = response_for_schema(schema_name)
        data = json.dumps(
            {
                "id": f"fake-{schema_name}",
                "object": "chat.completion",
                "choices": [
                    {
                        "index": 0,
                        "message": {"role": "assistant", "content": json.dumps(payload, ensure_ascii=False)},
                        "finish_reason": "stop",
                    }
                ],
            },
            ensure_ascii=False,
        ).encode("utf-8")
        self.send_response(200)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(data)))
        self.end_headers()
        self.wfile.write(data)

    def log_message(self, fmt, *args):
        return


def response_for_schema(schema_name):
    if schema_name == "task_profile_agent":
        return {
            "task_type": "真实模型教育场景审议",
            "industry": "教育",
            "user_goal": "形成弱硬件学校可落地的英语 AI 教学方案",
            "decision_type": "教学改革方案评审",
            "risk_level": "high",
            "required_perspectives": ["教育行业专家", "教学设计", "学习评估", "学校落地", "数据隐私"],
            "missing_information": ["学校网络条件", "教师 AI 能力"],
            "clarifying_questions": ["学校是否允许学生使用个人设备？"],
            "evidence_needs": ["课堂流程", "评价指标"],
            "assumptions": ["硬件条件弱是关键约束"],
            "constraints": ["不能依赖高配终端"],
        }
    if schema_name == "agent_selector_agent":
        return {
            "panel_summary": "模型选择教育、教学设计、学习评估、学校落地和隐私相关 Agent。",
            "selected_agent_ids": [
                "agent-instructional-designer",
                "agent-learning-assessment",
                "agent-school-deployment-advisor",
                "agent-data-privacy-officer",
                "agent-reality-checker",
            ],
            "excluded_agents": [{"agent_id": "agent-ui-designer", "reason": "当前不是视觉设计任务。"}],
            "missing_perspectives": [],
            "selection_notes": ["阵容围绕真实学校落地，不围绕演示 UI。"],
        }
    if schema_name == "agent_position_event":
        return {
            "claim": "需要从课堂目标、低硬件约束和教师工作流切入，而不是直接堆 AI 功能。",
            "rationale": "教育场景首先要证明教师能用、学生能完成、学校能管理。",
            "assumptions": ["学校硬件条件较弱"],
            "risks": ["教师负担增加", "效果不可量化"],
            "evidence_refs": ["task_profile"],
            "confidence_level": "medium_high",
            "challenge_requested": True,
        }
    if schema_name == "critic_agent":
        return {
            "target_event_id": "provider-will-be-overwritten",
            "challenge_type": "implementation_gap",
            "severity": "high",
            "challenge_text": "方案必须说明低硬件条件下如何采集、反馈和评价，否则不可落地。",
            "required_response": True,
            "evidence_refs": ["task_profile"],
        }
    if schema_name == "agent_response_revision":
        return {
            "challenge_id": "provider-will-be-overwritten",
            "response_type": "accepted",
            "response_text": "接受质疑，方案改为先做低硬件课堂闭环。",
            "accepted_changes": ["增加低硬件约束", "补充评价指标"],
            "remaining_disagreement": "仍需学校确认设备和网络。",
            "revised_claim": "先围绕教师备课、课堂练习和形成性反馈做低硬件闭环。",
        }
    if schema_name == "consensus_synthesizer":
        return {
            "consensus_type": "conditional_consensus",
            "statement": "可以推进，但必须以低硬件课堂闭环、教师工作流和学习评估为主线。",
            "supporting_event_ids": [],
            "confidence_level": "medium_high",
            "remaining_risks": ["学校设备条件未确认"],
            "disagreements": ["是否需要学生端设备仍未确定"],
            "next_actions": ["确认学校设备", "设计一节课试点", "定义评价指标"],
        }
    if schema_name == "report_synthesizer":
        return {
            "title": "教育场景圆桌评审报告",
            "executive_summary": "AI 英语教学应从低硬件课堂闭环切入。",
            "quality_score": "7/10",
            "fatal_flaws": [],
            "high_risk_gaps": ["学校设备和网络条件未确认"],
            "recommendations": ["先做教师端和课堂投屏/纸笔结合方案"],
            "next_actions": ["跑一节课试点"],
            "markdown": "# 圆桌评审报告\n\n## 一句话结论\n\n可以推进低硬件英语教学 AI 方案，但必须先试点验证。\n\n## 关键质疑\n\n低硬件约束未确认。\n",
        }
    return {}


def main():
    provider = ThreadingHTTPServer(("127.0.0.1", 0), FakeProvider)
    provider_port = provider.server_address[1]
    thread = threading.Thread(target=provider.serve_forever, daemon=True)
    thread.start()

    with tempfile.TemporaryDirectory() as tmp:
        orchestrator_port = find_free_port()
        env = os.environ.copy()
        env.update(
            {
                "AI_RUNTIME": "openai",
                "ORCHESTRATOR_HOST": "127.0.0.1",
                "ORCHESTRATOR_PORT": str(orchestrator_port),
                "DATA_DIR": tmp,
                "OPENAI_BASE_URL": f"http://127.0.0.1:{provider_port}",
                "OPENAI_API_MODE": "chat_completions",
                "OPENAI_API_KEY": "test-compatible-key",
                "OPENAI_MODEL": "qwen-compatible-test",
                "OPENAI_TIMEOUT_MS": "5000",
                "OPENAI_MAX_RETRIES": "0",
            }
        )
        proc = subprocess.Popen(
            [sys.executable, "ai-orchestrator-python/src/orchestrator.py"],
            cwd=Path(__file__).resolve().parents[2],
            env=env,
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            text=True,
        )
        try:
            wait_for(f"http://127.0.0.1:{orchestrator_port}/health", proc)
            created = post(
                orchestrator_port,
                "/api/sessions",
                {
                    "problem": "如何将 AI 赋能到英语教学当中，尤其是硬件条件比较弱的初中学校？",
                    "background": "初中、英语、论文。",
                    "targetOutput": "教育场景报告",
                },
            )
            if created["session"]["task_profile"]["runtime_mode"] != "openai":
                raise AssertionError("Task profile did not use openai runtime.")
            result = post(orchestrator_port, f"/api/sessions/{created['session']['session_id']}/run", {})
            if result["session"]["status"] != "complete":
                raise AssertionError("Roundtable did not complete.")
            schemas = {item["schema_name"] for item in provider_calls}
            expected = {"task_profile_agent", "agent_selector_agent", "agent_position_event", "critic_agent", "agent_response_revision", "consensus_synthesizer", "report_synthesizer"}
            missing = expected - schemas
            if missing:
                raise AssertionError(f"Provider calls missing: {missing}")
            if not all(item["authorization"] == "Bearer test-compatible-key" for item in provider_calls):
                raise AssertionError("Authorization header missing.")
            if not all(item["response_format"] == "json_object" for item in provider_calls):
                raise AssertionError("JSON response format missing.")
            print("Python OpenAI-compatible smoke passed:", created["session"]["session_id"], len(provider_calls), "provider calls")
        finally:
            proc.terminate()
            try:
                proc.wait(timeout=2)
            except subprocess.TimeoutExpired:
                proc.kill()
            provider.shutdown()


def find_free_port():
    server = ThreadingHTTPServer(("127.0.0.1", 0), BaseHTTPRequestHandler)
    port = server.server_address[1]
    server.server_close()
    return port


def wait_for(url, proc):
    deadline = time.time() + 10
    while time.time() < deadline:
        if proc.poll() is not None:
            output = proc.stdout.read() if proc.stdout else ""
            raise RuntimeError(f"orchestrator exited early:\n{output}")
        try:
            with urllib.request.urlopen(url, timeout=0.5) as response:
                if response.status == 200:
                    return
        except Exception:
            time.sleep(0.2)
    output = proc.stdout.read() if proc.stdout else ""
    raise TimeoutError(f"orchestrator did not start:\n{output}")


def post(port, path, body):
    request = urllib.request.Request(
        f"http://127.0.0.1:{port}{path}",
        data=json.dumps(body, ensure_ascii=False).encode("utf-8"),
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    with urllib.request.urlopen(request, timeout=15) as response:
        return json.loads(response.read().decode("utf-8"))


if __name__ == "__main__":
    main()

File: app/backend/test/openai-compatible-smoke.mjs
import { spawn } from "node:child_process";
import { once } from "node:events";
import fs from "node:fs";
import http from "node:http";
import path from "node:path";

const backendPort = 8878;
const testDataDir = path.resolve("backend/data-test/openai-compatible");
fs.rmSync(testDataDir, { recursive: true, force: true });

const providerCalls = [];
const fakeProvider = http.createServer(async (req, res) => {
  if (req.method !== "POST" || req.url !== "/chat/completions") {
    res.writeHead(404, { "Content-Type": "application/json" });
    res.end(JSON.stringify({ error: "not_found" }));
    return;
  }

  const chunks = [];
  for await (const chunk of req) chunks.push(chunk);
  const body = JSON.parse(Buffer.concat(chunks).toString("utf8"));
  const systemPrompt = body.messages?.find((message) => message.role === "system")?.content || "";
  const schemaName = systemPrompt.match(/schema_name=([^,，\s]+)/)?.[1] || "unknown";
  providerCalls.push({
    schemaName,
    authorization: req.headers.authorization || "",
    responseFormat: body.response_format?.type,
    model: body.model
  });

  const payload = responseForSchema(schemaName);
  res.writeHead(200, { "Content-Type": "application/json" });
  res.end(JSON.stringify({
    id: `fake-${schemaName}`,
    object: "chat.completion",
    choices: [
      {
        index: 0,
        message: {
          role: "assistant",
          content: JSON.stringify(payload)
        },
        finish_reason: "stop"
      }
    ]
  }));
});

await new Promise((resolve) => fakeProvider.listen(0, "127.0.0.1", resolve));
const providerPort = fakeProvider.address().port;

const child = spawn(process.execPath, ["backend/src/server.mjs"], {
  cwd: new URL("../..", import.meta.url),
  env: {
    ...process.env,
    BACKEND_PORT: String(backendPort),
    APP_BASE_PATH: "/agent-roundtable-studio",
    DATA_DIR: "./backend/data-test/openai-compatible",
    AI_RUNTIME: "openai",
    OPENAI_BASE_URL: `http://127.0.0.1:${providerPort}`,
    OPENAI_API_MODE: "chat_completions",
    OPENAI_API_KEY: "test-compatible-token",
    OPENAI_MODEL: "qwen-compatible-test",
    OPENAI_TIMEOUT_MS: "5000",
    OPENAI_MAX_RETRIES: "0"
  },
  stdio: ["ignore", "pipe", "pipe"]
});

let output = "";
child.stdout.on("data", (chunk) => {
  output += chunk.toString();
});
child.stderr.on("data", (chunk) => {
  output += chunk.toString();
});

async function waitForServer() {
  const deadline = Date.now() + 8000;
  while (Date.now() < deadline) {
    try {
      const response = await fetch(`http://127.0.0.1:${backendPort}/health`);
      if (response.ok) return;
    } catch {
      await new Promise((resolve) => setTimeout(resolve, 150));
    }
  }
  throw new Error(`Backend did not start. Output:\n${output}`);
}

async function post(route, body = {}) {
  const response = await fetch(`http://127.0.0.1:${backendPort}${route}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body)
  });
  if (!response.ok) {
    throw new Error(`${route} failed: ${response.status} ${await response.text()}`);
  }
  return response.json();
}

async function get(route) {
  const response = await fetch(`http://127.0.0.1:${backendPort}${route}`);
  if (!response.ok) {
    throw new Error(`${route} failed: ${response.status} ${await response.text()}`);
  }
  return response.json();
}

function responseForSchema(schemaName) {
  if (schemaName === "task_profile_enrichment") {
    return {
      task_type: "第三方兼容 API 增强任务画像",
      risk_level: "high",
      required_perspectives: ["产品", "财务审计", "医疗合规"],
      evidence_needs: ["模型兼容调用记录", "自动补位 Agent 记录"],
      assumptions: ["第三方兼容服务返回 JSON object"],
      constraints: ["不得将真实 API Key 写入仓库"]
    };
  }

  if (schemaName === "agent_panel_review") {
    return {
      panel_summary: "第三方兼容 API 已复核阵容，建议保留自动补位 Agent 并标记为个人试用。",
      risks: ["财务审计和医疗合规需要明确证据边界"],
      missing_expertise: [],
      selection_notes: ["自动补位 Agent 只能用于当前用户任务"]
    };
  }

  if (schemaName === "roundtable_report_polish") {
    return {
      executive_summary: "第三方兼容 API 已复核报告，当前结论可作为条件性评审输出。",
      quality_score: "7/10",
      high_risk_gaps: ["真实供应商模型稳定性仍需上线前压测"],
      next_actions: ["用真实供应商 Key 在服务器环境变量中完成一次受控验证"]
    };
  }

  return {};
}

try {
  await waitForServer();

  const health = await get("/health");
  if (!health.openaiConfigured || health.runtime !== "openai") {
    throw new Error("Backend did not start in configured openai runtime mode.");
  }

  const created = await post("/api/sessions", {
    problem: "验证阿里云百炼 OpenAI-compatible API 接入：系统应自动组建圆桌，不足时生成个人 Agent，并展示第三方 API 阵容复核。",
    background: "本测试使用本地假兼容服务，不访问真实供应商，不包含真实 Key。",
    targetOutput: "兼容 API 集成验收报告"
  });

  if (created.session.task_profile.ai_enrichment?.status !== "applied") {
    throw new Error("Task profile AI enrichment was not applied.");
  }
  if (!created.session.recommendation?.ai_review || created.session.recommendation.ai_review.status !== "applied") {
    throw new Error("Agent panel AI review was not applied.");
  }
  if (!created.session.recommendation.assembly_trace.some((item) => item.stage === "ai_panel_review")) {
    throw new Error("AI panel review stage missing from assembly trace.");
  }
  if (!created.session.recommendation.generated_agents?.length) {
    throw new Error("Expected generated personal agents for third-party enriched perspectives.");
  }

  const result = await post(`/api/sessions/${created.session.session_id}/run`);
  if (result.session.status !== "complete") {
    throw new Error(`Expected complete session, got ${result.session.status}`);
  }
  if (result.artifact.ai_review?.status !== "applied") {
    throw new Error("Roundtable report AI review was not applied.");
  }
  if (!result.artifact.markdown.includes("第三方 API 复核摘要")) {
    throw new Error("Final report does not include third-party API appendix.");
  }

  const schemas = providerCalls.map((call) => call.schemaName);
  for (const expected of ["task_profile_enrichment", "agent_panel_review", "roundtable_report_polish"]) {
    if (!schemas.includes(expected)) throw new Error(`Provider call missing: ${expected}`);
  }
  if (!providerCalls.every((call) => call.responseFormat === "json_object")) {
    throw new Error("Chat completions requests did not require JSON object responses.");
  }
  if (!providerCalls.every((call) => call.authorization === "Bearer test-compatible-token")) {
    throw new Error("Authorization header was not sent to compatible provider.");
  }

  console.log("OpenAI-compatible smoke passed:", created.session.session_id, `${providerCalls.length} provider calls`);
} finally {
  child.kill();
  fakeProvider.close();
  await Promise.race([
    once(child, "exit"),
    new Promise((resolve) => setTimeout(resolve, 1000))
  ]);
}
